### ISSUE TITLE
Make code work on no_std

### ISSUE THREAD

**Issue by TheButlah** [kitlith/vqf-cxx#3]
We want to try to use this filter in the [rust firmware](https://github.com/SlimeVR/SlimeVR-Rust/tree/main/firmware) we are writing for slimevr. Having it working on `#![no_std]` and ideally even without `alloc` would be awesome.

**Comment by kitlith** [comment]
so, my code doesn't allocate, and the only use of std should be easily replaced with core. However, the code that autocxx generates may be a different story. ~~I can't tell whether it would also just be fine with core or not.~~ Am checking if there's a way to put it into a no-std mode or if we need to submit a patch.

EDIT: generated code wants std::pin::Pin and std::mem::MaybeUninit, both of which I think are in core. It also wants std::String, which I don't think we need.

inb4 there are other deps that are hidden somehow.

**Comment by kitlith** [comment]
I've made a hacked up autocxx tree, we should try to upstream parts of it: https://github.com/kitlith/autocxx/tree/no_std

I have also pushed a no_std branch to this repository that depends on the hacked up autocxx tree, and massages the dependencies so it successfully performs a `check` against a no_std target. However, it still wants to link to libc++ or libstdc++. That might be a blocker for you.

**Comment by TheButlah** [comment]
yeah, linking to libc/++ would be a nono here. Does the original C code use standard library stuff? Either way, upstreaming no_std support for autocxx would be pretty cool.

**Comment by kitlith** [comment]
Found out autocxx actually has an option to disable all the usually helpful codegen, so that's even less of an issue than i originally thought.

However, cxx, even, if we remove the bits about unique_ptr and string and such still wants the c++ stdlib parts that are equivalent to core (i.e. language support for placement new, type traits) to enforce its safety invariants.

This is fine if you've built a version of the c++ stdlib for your target (especially if you have a toolchain to go with it). I'm not sure there's a good way to integrate that into cargo atm, though.

GNU's libstdc++ seems to be "yeah you build this at the same time as the compiler" but they do explicitly support freestanding c++. Maybe there's a way to build it on its own, idk.

LLVM's libcxx/libc++ has flags to disable parts of the stdlib, and an ongoing patch for c++23's freestanding definition. (current version here: https://reviews.llvm.org/D140691, previous discussion here: https://reviews.llvm.org/D135911) I've not looked very closely at what the current flags would let us disable, so maybe it's not necessary.

I think it would be *really cool* if we could build libc++ as a crate, enable lto, the whole nine yards. But it's somewhat moot for *this* crate since y'all are porting vqf to rust anyway.

I might still investigate anyway. would be neat if libc++ compile times aren't hell.
